README
                                  INTRODUCTION

FLASH (Fast Length Adjustment of SHort reads) is an accurate and fast tool
to merge paired-end reads that were generated from DNA fragments whose
lengths are shorter than twice the length of reads.  Merged read pairs result
in unpaired longer reads, which are generally more desired in genome
assembly and genome analysis processes.

Briefly, the FLASH algorithm considers all possible overlaps at or above a
minimum length between the reads in a pair and chooses the overlap that
results in the lowest mismatch density (proportion of mismatched bases in
the overlapped region).  Ties between multiple overlaps are broken by
considering quality scores at mismatch sites.  When building the merged
sequence, FLASH computes a consensus sequence in the overlapped region.
More details can be found in the original publication
(http://bioinformatics.oxfordjournals.org/content/27/21/2957.full).

Limitations of FLASH include:
   - FLASH cannot merge paired-end reads that do not overlap.
   - FLASH cannot merge read pairs that have an outward orientation, either
     due to being "jumping" reads or due to excessive trimming.
   - FLASH is not designed for data that has a significant amount of indel
     errors (such as Sanger sequencing data).  It is best suited for Illumina
     data.

                                  INSTALLATION

On UNIX-compatible systems, including GNU/Linux and Mac OS X, you must compile
FLASH from source.  The only dependency, other than functions that are expected
to be available in the C library, is the zlib data compression library.  To
install FLASH, download the tarball, untar it, and compile the code using the
provided Makefile:

    $ tar xzf FLASH-1.2.9.tar.gz
    $ cd FLASH-1.2.9
    $ make

The executable file that is produced is named 'flash'.  To run it from the
command line you must copy it to a location on your $PATH variable, or else run
it with a path including a directory, such as "./flash".

FLASH also runs on Windows, and you can compile it on Windows using MinGW.
However, for convenience you may instead download a standalone Windows binary
from the SourceForge page (https://sourceforge.net/projects/flashpage/).

                                     USAGE

Please compile FLASH and run `flash --help' to see command-line usage
information and information about input/output files.

                                 MULTITHREADING

By default, FLASH uses multiple threads.  There are "combiner" threads that do
the actual read combining, as well as up to 5 threads that are used for I/O (up
to 2 readers, up to 3 writers).  The default number of combiner threads is the
number of processors; however, it can be adjusted with the -t  option (long
option: --threads).

When multiple combiner threads are used, the order of the combined and
uncombined reads in the output files will be nondeterministic.  If you need to
enforce that the output reads appear in the same order as the input, you must
specify --threads=1.

                                  PERFORMANCE

Since the FLASH algorithm considers each read pair independently, FLASH will, by
default, process read pairs in parallel.  FLASH v1.2.9 and later also make use
of vector instructions available on modern x86 CPUs.  Consequently, FLASH works
quite fast, even with low-cost computing resources.  As an example, we ran FLASH
v1.2.9 on a laptop with a dual-core 2.3 GHz AMD x86_64 processor and it
processed one million 101-bp read pairs in 11.6 seconds with the default
parameters.  Less than 2 MB of memory was used.  Actual timing results will
vary, but they will depend primarily on the number of CPUs available, the speed
of each CPU, and on the I/O speed of reading the input files and writing the
output files.  FLASH is designed to be scalable to dozens of processors,
although its speed may be limited by I/O in such cases.

                                   ACCURACY

With reads' error rate of 1% or less, FLASH processes over 99% of read pairs
correctly.  With error rate of 2%, FLASH processes over 98% of read pairs
correctly when default parameters are used. With more aggressive parameters
(i.e., -x 0.35), FLASH processes over 90% of read pairs correctly even when the
error rate is 5%.

                                  PUBLICATION

Title:   FLASH: fast length adjustment of short reads to improve genome assemblies
Authors: Tanja Magoč and Steven L. Salzberg
URL:     http://bioinformatics.oxfordjournals.org/content/27/21/2957.full

                                    LICENSE

FLASH is released under the GNU General Public License Version 3 or later (see
COPYING).

                          COMMENTS/QUESTIONS/REQUESTS

Send an e-mail to flash.comment@gmail.com

Other versions are available from the SourceForge page:

https://sourceforge.net/projects/flashpage/


NEWS
Version 1.2.9
-------------------
- The core algorithm of FLASH has been optimized using SSE and SSE2
  instructions.  On x86_64 CPUs it now can run over twice as fast.
- The help output of FLASH has been improved.
- Added supported for tab-delimited input and output.
- FLASH now displays more informative error messages in some cases.
- Reorganized some of the code so that it's easier to modify.
- Reading the data from standard input now works correctly.
- With --to-stdout specified, informational messages are now written to standard
  error rather than omitted.

Version 1.2.8
-------------------
- FLASH's default behavior when assigning quality scores to mismatch sites in
  the overlap region of combined reads has been improved.  See the documentation
  for the new --cap-mismatch-quals option (which can be provided to request the
  old behavior if desired) for more information about this change.

Version 1.2.7
-------------------
- FLASH is now supported on Windows.  See the README.
- Minor bug-fixes regarding printing of error messages.
- Fixed a bug where the time that FLASH took to run may have been mis-calculated
  on 32-bit systems in an unlikely case.  This only affected the printed
  informational output.
- Updated PERFORMANCE section in README to be accurate for the current version
  of FLASH.

Version 1.2.6
-------------------
- FLASH now works on files larger than 2GiB on 32-bit operating systems, unless
  the system zlib library does not support such files.  64-bit operating systems
  are unaffected as they already supported large files by default.

Version 1.2.5
-------------------
- FLASH will now compile and run on Mac OS X.

Version 1.2.4
-------------------
- Support for piping the output of FLASH through arbitrary compression programs
  has been added via the --compress-prog option.  The --compress-prog-args
  and/or --output-suffix options also may be useful.
- FLASH with the --interleaved option will now use an interleaved output format
  in addition to assuming interleaved input.  Use --interleaved-input or
  --interleaved-output to specify only interleaved input or only interleaved
  output, respectively.
- The entirely single-threaded configuration of FLASH is no longer supported.
  You can still use --threads=1 to use only 1 combiner thread, however.

Version 1.2.3
-------------------
- Support for interleaved reads (-I, --interleaved), which may be read from one
  FASTQ file or piped in from stdin
- Fix --read-len option
- Fix default --fragment-len-stddev option
- Fix reading FASTQ files with embedded zero bytes
- Fix memory leaks
- Added license file

Version 1.2.2
-------------------
- Fix linking problem in Makefile

Version 1.2.1
-------------------
- There is no longer any read length or tag length limit, although you may still
specify the --read-len parameter for the purpose of automatically computing the
maximum overlap.
- FLASH now prints some informational output while it's running, unless running
with --quiet or --to-stdout.

Version 1.2
-------------------

- Support for writing extended fragments directly to standard output.
- Support for writing compressed output files.
- Support for multithreading.  There are multiple I/O threads as well as
multiple combiner threads.  Speedup of 3.6 times has been observed.  Compiling
without support for multithreading is still supported.

Version 1.1
-------------------
Rewrote the entire program.
- Removed at least 3 instances of undefined behavior (such as using
uninitialized variables).
- Performance improvements. Neglecting I/O, the program now runs over twice as
fast.  This improvement is not due to a change in algorithm; I expect it is due
mainly to the rewritten align_position() function.
- Use getopt_long() for option parsing.  Now, long options are accepted, and
options do not have to be given after the mate pair FASTQ files.
- Removed many fixed-size buffers.  This saves memory, removes many buffer
overflow problems, and removes the READ_LENGTH parameter.  Note: the maximum tag
length is still hard-coded as 1024 bytes.
- Much more error checking, so the program will cleanly fail if something goes
wrong.
- Use a Makefile instead of a shell script.
- The '.histogram' file is written a bit more narrowly (80 columns), but other
than that, the output of the program should be the same (unless it was affected
by any of the cases of undefined behavior I found).
- Some changes to the MANUAL, although I didn't feel like changing the
information that is redundant with the program help.
- Note: I am calling this version 1.1, but feel free to call it something else.
The version number is also listed in the version() function in flash.c.

Version v1.0.3
--------------
1. FLASH allows reads files to be in gzip format.
2. FLASH builds a histogram of fragment lengths produced by merged read pairs.
3. FLASH now allows for reads longer than 170bp to be processed (READ_LENGTH is utilities.h file still needs to be changed manually to the read length if it is greater than 170).
4. If original reads' ID contains barcode, the barcode is kept as a part of the ID in the output fragments.



Version v1.0.2
--------------
Fixed bugs:
1. In combine function in combineReads.c: modified selection of the base when quality values of bases are the same. Instead of random selection of the base, first check whether any base is "N" to avoid selection of "N" if not necessary.

Added options for easier handling of the program:
1. Added error messages in case no input files and no parameters are specified.
2. Added the parameter to print 'help' for running the program.
3. Added the option to specify prefix of the output files.
4. Added the option to specify directory of the output files.

Functional changes:
1. Changed read length to 170.
2. Automatic calculation of maximum overlap length given average read length, average fragment length, and fragment standard deviation.


Version v1.0.1
--------------
Fixed bugs:

extend.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "complementReverse.h"
#include "combineReads.h"
#include "utilities.h"
#include <zlib.h>

/*Usage: flash <mates1.fastq> <mates2.fastq> [-m minOverlap] [-M maxOverlap] [-x mismatchRatio] [-p phredOffset] [-o prefixOfOutputFiles] [-d pathToDirectory] [-r readLength] [-f fragmentlength] [-s fragmentStandardDeviation]*/  
/*Try 'flash -h' for more information.*/

int main(int argc, char **argv)
{ 
  int result,i,compareStr,checkPar;
  int start,j,k,dir,stop,length,position,max_length,max_hist;
  int min_overlap,max_overlap,max_overlap_flag,phred_offset,read_length,frag_length;
  int len[FINAL_READ_LENGTH];
  float max_mismatch, frag_std;
  char prefix[30]="out";
  char read[FINAL_READ_LENGTH];
  char outFile[60],notCombined1[60],notCombined2[60],hist[60],histogram[60];
  char directory[200]=".";
  char temp[20][200];
  char tempPath[200]="";
  char blank[1]="";
  char id[READ_LENGTH];
  char file1[4][READ_LENGTH],file2[4][READ_LENGTH],finalFile[4][FINAL_READ_LENGTH];
  FILE *outFile_p,*notCombined1_p,*notCombined2_p,*hist_p,*histogram_p;
  gzFile mates1_g,mates2_g;

  struct stat st;

  phred_offset=33;
  max_overlap=70; /*max overlap value used to calculate mismatch score*/
  max_overlap_flag=0;
  max_mismatch=0.25; /*max ratio=(number of mismatches)/(overlap length)*/
  min_overlap=10;
  read_length=100;
  frag_length=180;
  frag_std=20;

  if(argc==2){
    if((argv[1][0]=='-') && (argv[1][1]=='h')){
      printHelp();
      return(1);
    }
    else{
      printf("At least one input file is not specified.\n");
      printForHelp();
      return(1);
    }
  }
  else{
    if(argc<3){
      printf("At least one input file is not specified.\n");
      printForHelp();
      return(1);
    }
  }

  if(argv[1][0]=='-'){
    printf("One of two input files is not specified or its name starts with '-'. The names of these files cannot start with '-'.\n");
    printForHelp();
    return(1);
  }
  mates1_g=gzopen(argv[1],"r");
  if(mates1_g == NULL){
    printf("The first specified file cannot be opened, or is empty.\n");
    printForHelp();
    return(1);
  }
  if(argv[2][0]=='-'){
    printf("One of two input files is not specified or its name starts with '-'. The names of these files cannot start with '-'.\n");
    printForHelp();
    return(1);
  }
  mates2_g=gzopen(argv[2],"r");
  if(mates2_g == NULL){
    printf("The second specified file cannot be opened, or is empty.\n");
    printForHelp();
    return(1);
  }

  for(i=3;i<argc;i++){
    if(argv[i][0]=='-'){
      switch(argv[i][1]){
        case 'm':
          i++;
	  checkPar=checkParity('m',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    min_overlap=atoi(argv[i]);
	    if(min_overlap<1){
	      printf("Minimum overlap should be positive.\n");
	      printForHelp();
	      return(1);
	    }
	    break;
	  }
        case 'M':
          i++;
	  checkPar=checkParity('M',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    max_overlap=atoi(argv[i]);
	    max_overlap_flag=1;
	    if(max_overlap<1){
	      printf("Maximum overlap should be positive.\n");
	      printForHelp();
	      return(1);
	    }
	    break;
	  }
        case 'x':
          i++;
	  checkPar=checkParity('x',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    max_mismatch=atof(argv[i]);
	    if((max_mismatch<0) || (max_mismatch>1)){
	      printf("Max mismatch rate has to be in the interval [0,1].\n");
	      printForHelp();
	      return(1);
	    }
	    break;
	  }
        case 'o':
	  i++;
	  checkPar-checkParity('o',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    strcpy(prefix,argv[i]);
	    break;
	  }
        case 'd':
	  i++;
	  checkPar=checkParity('d',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    strcpy(directory,argv[i]);
	    break;
	  }
        case 'p':
          i++;
	  checkPar=checkParity('p',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    phred_offset=atoi(argv[i]);
	    if((phred_offset!=33) && (phred_offset!=64)){
	      printf("WARNING: Phred offset is usually either 64 (for earlier Illumina data) or 33 (for Sanger and later Illumina data).\n"); 
	      printForHelp();
	    }
	    break;
	  }
        case 'r':
	  i++;
	  checkPar=checkParity('r',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    read_length=atoi(argv[i]);
	    if((read_length<=0) || (read_length>READ_LENGTH)){
	      printf("Read length has to be in the range (0,170]. If the reads are longer than 170, please change the constant value for READ_LENGTH in utilities.h\n");
	      printForHelp();
	      return(1);
	    }
	    break;
	  }
        case 'f':
	  i++;
	  checkPar=checkParity('f',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    frag_length=atoi(argv[i]);
	    if(frag_length<=0){
	      printf("Fragment length must be positive.\n");
	      printForHelp();
	      return(1);
	    }
	    break;
	  }
        case 's':
	  i++;
	  checkPar=checkParity('s',i,argc);
	  if(checkPar==1){
	    return(1);
	  }
	  else{
	    frag_std=atof(argv[i]);
	    if(frag_std<=0){
	      printf("Standard deviation of fragments must be positive.\n");
	      printForHelp();
	      return(1);
	    }
	    break;
	  }
        case 'h':
	  printHelp();
	  return(1);
	  break;
        default:
          printf("Unknown option. %s\n.",argv[i]);
          printf("Usage: flash <mates1.fastq> <mates2.fastq> [-m minOverlap] [-M maxOverlap] [-x mismatchRatio] [-p phredOffset] [-o prefixOfOutputFiles] [-d pathToOutputDirectory] [-r readLength] [-f fragmentLength] [-s fragmentStandardDeviation]\n");
	  printForHelp();
          return(1);
      }
    }
    else{  /*else of if(argv[i][0]=='-'*/
      printf("Unknown input parameter %s.\n", argv[i]);
      printForHelp();
    }
  }

  if(max_overlap_flag==0){
    if((read_length!=100) || (frag_length!=180) || (frag_std!=20)){
      max_overlap=(int)(2*read_length-frag_length+2.5*frag_std);
    }
  }

  if(directory!="."){
    j=0;
    strcpy(temp[0],directory);
    start=0;
    dir=strlen(directory);
    for(i=0;i<dir;i++){
      if((directory[i]=='/') | (directory[i]=='\\')){
	strncpy(temp[j],directory+start,i-start);
	temp[j][i]='\0';
	j++;
	start=i+1;
      }
    }
    strncpy(temp[j],directory+start,i-start);
    temp[j][i]='\0';
    j++;
    stop=0;

    /*check if the path is given as absolute path*/
    if((directory[0]=='/') || (directory[0]=='\\')) { /*absolute path is provided*/
      /*check which part of absolute path exists*/
      i=1;
      strcpy(tempPath,blank);
      strcat(tempPath,"/");
      strcat(tempPath,temp[1]);
      while(stat(tempPath,&st)==0){
	i++;
	strcat(tempPath,"/");
	strcat(tempPath,temp[i]);
      }
      stop=i;
      i=j;
    }

      /*change directory to the latest existing stage of absolute path*/
    if(stop==1){
      strcpy(tempPath,"/");
      printf("WARNING: You are trying to create a directory at the root. If you receive a 'Segmentatoin fault' message, you most probably do not have right to create a directory at root. \n");
    }
    else{
      strcpy(tempPath,blank);
      for(i=1;i<stop;i++){
	strcat(tempPath,"/");
	strcat(tempPath,temp[i]);
      }
    }
    chdir(tempPath);

    /*create directories for output*/
    for(i=stop;i<j;i++){
      mkdir(temp[i], S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
      chdir(temp[i]);
    }
  }

  strcpy(notCombined1,prefix);
  strcat(notCombined1,".notCombined_1.fastq");
  strcpy(notCombined2,prefix);
  strcat(notCombined2,".notCombined_2.fastq");
  strcpy(outFile,prefix);
  strcat(outFile,".extendedFrags.fastq");
  notCombined1_p=fopen(notCombined1,"w");
  notCombined2_p=fopen(notCombined2,"w");
  outFile_p=fopen(outFile,"w");

  gzgets(mates1_g,file1[0],READ_LENGTH);
  removeEndBlanks(file1[0]);
  while(!gzeof(mates1_g)){
    for(i=1;i<4;i++){
      gzgets(mates1_g,file1[i],READ_LENGTH);
      removeEndBlanks(file1[i]);
    }
    for(i=0;i<4;i++){
      gzgets(mates2_g,file2[i],READ_LENGTH);
      removeEndBlanks(file2[i]);
    }
    uppercase(file1[1]);
    compRev(file2[1]);
    reverse(file2[3]);
    idSet(file1[0],file2[0],id);
    result=combine(id,file1[1],file2[1],file1[3],file2[3],finalFile[1],finalFile[3],min_overlap,max_overlap,max_mismatch,phred_offset);
    if(result){
      fprintf(outFile_p,"%s\n%s\n+\n%s\n",id,finalFile[1],finalFile[3]);
    }
    else{
      fprintf(notCombined1_p,"%s\n%s\n+\n%s\n",file1[0],file1[1],file1[3]);
      compRev(file2[1]);
      reverse(file2[3]);
      fprintf(notCombined2_p,"%s\n%s\n+\n%s\n",file2[0],file2[1],file2[3]);
    }
    gzgets(mates1_g,file1[0],READ_LENGTH); /*get id of the next read. Do it here because if it is EOF, it will not go into while-loop*/
      removeEndBlanks(file1[0]);
  }

  gzclose(mates1_g);
  gzclose(mates2_g);
  fclose(outFile_p);
  fclose(notCombined1_p);
  fclose(notCombined2_p);

  /*Make histogram of lengths of merged reads*/
  for(i=read_length;i<(FINAL_READ_LENGTH+1);i++){
    len[i]=0;
  }
  max_length=0;
  outFile_p=fopen(outFile,"r");
  fgets(read,FINAL_READ_LENGTH,outFile_p);
  while(!feof(outFile_p)){
      fgets(read,FINAL_READ_LENGTH,outFile_p);
      length=readLength(read);
    if(length>max_length){
      max_length=length;
    }
    len[length]++;
    for(i=0;i<3;i++){
      fgets(read,FINAL_READ_LENGTH,outFile_p);
    }
  }
  fclose(outFile_p);

  max_hist=0;
  strcpy(hist,prefix);
  strcat(hist,".hist");
  hist_p=fopen(hist,"w");
  for(i=read_length;i<=max_length;i++){
    fprintf(hist_p,"%d\t%d\n",i,len[i]);
    if(len[i]>max_hist){
      max_hist=len[i];
    }
  }
  fclose(hist_p);

  j=max_hist/100;
  if(j==0){
    j=1;
  }
  strcpy(histogram,prefix);
  strcat(histogram,".histogram");
  histogram_p=fopen(histogram,"w");
  hist_p=fopen(hist,"r");
  fscanf(hist_p,"%d %d",&position,&length);
  while(!feof(hist_p)){
    if(length==0){
      k=0;
    }
    else{
      k=length/j;
      if(k==0){
	k=1;
      }
    }
    fprintf(histogram_p,"%d\t",position);
    for(i=0;i<k;i++){
      fprintf(histogram_p,"*");
    }
    fprintf(histogram_p,"\n");
    fscanf(hist_p,"%d %d",&position,&length);
  }
  fclose(hist_p);
  fclose(histogram_p);

  return(0);
}
